### DISCUSSION TITLE
useInfiniteQuery replies requests for all cached pages at once

### DISCUSSION THREAD

**Discussion by levani** [TanStack/query#1670] [Q&A]
I'm using `useInfiniteQuery` in a react native project to fetch data for `FlatList` component when the user scrolls to the bottom.

If I scroll down to, lets say page 5, then navigate to another screen and then return back, `useInfiniteQuery` sends requests for all five pages without the user even touching the `FlatList`. So just by opening the page I get 5 requests:

http://example.com/news?page=1
http://example.com/news?page=2
http://example.com/news?page=3
http://example.com/news?page=4
http://example.com/news?page=5

Is this how `useInfiniteQuery` supposed to work? If it is, then is there a way to fetch only the first page on page load?

**Comment by TkDodo** [comment]
Yes, it’s supposed to work that way. It will refetch all pages in the background one after the other. If you don’t that, you can reduce the `cacheTime`, but then you wind up with no data and a loading spinner, or maybe delete all pages except the first one via `queryClient.setQueryData`, though I’m not sure if that really works.

**Reply by levani** [reply]
Really curious why is that the default behavior... All pages should stay in cache until the `cacheTime` expires but why refetch them all at once? What's the real use case for this?

**Reply by TkDodo** [reply]
They are refetched because they are stale and you mount that component (refetchOnMount). You can also set a `staleTime`. If 5 pages are visible on the screen and are stale, 5 pages will be refetched. Why should it only be the first page? It would make the first page up to date, but not the 4 other pages.

**Reply by levani** [reply]
If they are visible yes they should be refetched but if they are not? Usually there is only the first page visible and when user scrolls only then is the next page's data necessary. What I expected is that `useInfiniteQuery` would display next page's data from cache and also fetch new data from server if it's stale, but only after I explicitly request that page using `fetchNextPage`. If the user never scrolls the list we end up sending x number of request to server but their data will never be used, isn't that so?

**Reply by TkDodo** [reply]
Yes, I see where you're coming from. The problem is that `fetchNextPage` would, in this case, fetch the 6th page once you call it, not the 2nd one. There are no params passed to `fetchNextPage`, because they come from `getNextPageParam`, which usually inspects the `lastPage` (the 5th one) and returns the next pointer. Otherwise, you would have to keep track of which page is the next page in user-land. With bi-directional infinite queries, I'm also not sure if the 0th page will always be the one you first see on the screen, so it gets into a really difficult problem space.

As of now, this is per design. Maybe we can come up with a more fine-grained refetching approach for infinite queries, but it has to cover all the bases :)

**Reply by mallchel** [reply]
@TkDodo Maybe I can pass some `reset` param when call `refetch` function from useInfiniteQuery to clear cache and get only the first page. I think it would have been appropriate in a case with large lists

**Reply by hirbod** [reply]
I just stumbled on the same issue and don't like the fact that it refetches all pages after I revist a screen (in sequence). It is killing the whole point of FlatList, at least in React Native.

I am calling the `remove` method now on unmount. I tried to use the 2nd refetchPage parameter but that did not remove the pages. Setting staleTime and cacheTime didn't change anything.

Now I need to figure out how I can `remove` but leave the first page.

```
  const { isLoading, data, hasNextPage, fetchNextPage, remove, isFetchingNextPage } = usePlaylists();
```

**Reply by TkDodo** [reply]
The reason why we refetch all pages is because it is the safest way to get to data consistency. If you have two pages in the cache and we only refetch the first one, if one entry was deleted on page1, the last entry on page1 (refetched) will be the same as the first entry on page2 (not refetched), so you'll have a duplicate.

If you want to remove all pages except the first one, you can do that with `setQueryData`:

```
setQueryData(['my-infinite-query'], (prev) => ({
  pages: prev.pages.slice(0,1),
  pageParams: prev.pageParms.slice(0,1),
})
```

imperative refetches, like `queryClient.invalidateQueries` do accept a `refetchPage` option where you can decide which pages to refetch and which not. But for automatic refetches, this does not exist.

**Reply by hirbod** [reply]
I get your point, but its a heavy performance impact with RN @TkDodo.
Your suggestion did not work for me unfortunately, it still keeps pages and pageParams in the cache and just changed the status from "fresh" to inactive. page and pageParams look the same:

<img width="386" alt="Bildschirmfoto 2022-03-02 um 18 19 36" src="https://user-images.githubusercontent.com/504909/156414085-acc063ca-9069-4d86-b586-9c4937657612.png">

**Reply by TkDodo** [reply]
setQueryData should work. Can you reproduce this in a codesandbox please

**Reply by hirbod** [reply]
I tried plenty of things. The only thing that worked changing with "setQueryData" was returning an empty object.

```tsx
      queryClient.setQueryData('playlists', (prev) => {
        console.log(prev); // this triggers
        return {
          pages: prev.pages.slice(0, 1),
          pageParams: prev.pageParms.slice(0, 1),
        }; // does absolutely nothing, as seen in my Flipper screenshot
      });
```

```tsx
      queryClient.setQueryData('playlists', (prev) => {
        return {} // this works and also is explorable via Flipper
      });
```

Do I have to set anything special? This is my Hook

```tsx
export default function usePlaylists() {
    return useInfiniteQuery<IPlayListPaginated, AxiosError>("playlists", ({ signal, pageParam = 1 }) => getPlaylists({ pageParam, signal }), {
        getNextPageParam: (lastPage) => {
            if (lastPage.links.next != null) {
                const nextPageParam = getUrlParameter(lastPage.links.next, "page");
                return nextPageParam;
            }
            return undefined;
        },
    });
}
```

**Reply by hirbod** [reply]
LOL... never trust code you've copied :D.... and I need some sleep.

I didn't realize the typo `pageParms` (which should be `pageParams`).
So weird it does not throw any error.. Looks like it gets catched internally.

Thank you very much @TkDodo - this is now working exactly how I need it.

**Reply by TkDodo** [reply]
Obligatory "Typescript would've caught that" 😅. Good to know that it works 👍

**Reply by hirbod** [reply]
I am usually typing most of the things, but prev is type "unknown" and I didn't bother to fix that for a "quick" test.

**Reply by mnhatwwpt** [reply]
Hi @hirbod , where did you use the 'setQueryData' function to keep only the first page ?

**Reply by CodeWithAlamin** [reply]
I followed your method (which is also showed in TanStack query official website) like this as you mentioned:
```
      queryClient.setQueryData('playlists', (prev) => {
        console.log(prev); // this triggers
        return {
          pages: prev.pages.slice(0, 1),
          pageParams: prev.pageParms.slice(0, 1),
        }; // does absolutely nothing, as seen in my Flipper screenshot
      });
```

But, the problem is, if i do that then TanStack query gets confused and keeps the "isFetching, isFetchingNextPage" etc states to "true". as a result if i go to another route and come back, these states becomes true which runs my loading spinner.

**Comment by cesarerodrigu3z** [comment]
Has this been resolved? is a very annoying problem, same here I am playing on react native application and the experience is not great

**Reply by TkDodo** [reply]
what behaviour would you expect please?

**Reply by auwtch** [reply]
I think a built-in solution to reset the cursor or page number after unmounting would solve this

**Reply by TkDodo** [reply]
> I think a built-in solution to reset the cursor or page number after unmounting would solve this

what do you mean with `resetting` ? A refetch always occurs from the first page, so it is “reset”. If you have 5 pages in the cache, do you want react-query to delete everything except the first page when you unmount a hook?

**Reply by hirbod** [reply]
Actually, that’s a good question. We’re manually doing this on pull-to-refresh in our apps, for example. If a user pulls to refresh, we remove all pages for that specific query but keep page 1 and refresh that. Anything else has proven to spam the server too much. Since we finally have bi-direction support, I still need to figure out how all of this works with maxPages.

Example:

```tsx
      trpcUtils.post.feed.setInfiniteData(feedQueryOptionsInput, (d) => {
        if (d?.pages.length) {
          return {
            pages: d.pages.slice(0, 1),
            pageParams: d.pageParams.slice(0, 1),
          }
        }
      })
      await refetch()
```

**Reply by TkDodo** [reply]
that’s totally fine, and I don’t think we need to provide more out of the box. It’s also documented here: https://tanstack.com/query/v5/docs/framework/react/guides/infinite-queries#keep-only-the-first-page

**Reply by hirbod** [reply]
I wish I didn’t have to repeat all this boilerplate code for all our infinite lists, to be honest. Would be great if refetch could accept an argument or something similar. Anyway, it's not too much effort though. Just need to make sure that the params are the same
